ufl/corealg/dag_traverser.py

```diff
@@ -0,0 +1,120 @@
+"""Base class for dag traversers."""
+
+from abc import ABC, abstractmethod
+from functools import singledispatchmethod, wraps
+
+from ufl.classes import Expr
+
+
+class DAGTraverser(ABC):
+    """Base class for dag traversers."""
+
+    def __init__(self, compress=True, vcache=None, rcache=None) -> None:
```

Comment:
I think `rcache` and `vcache` are very obscure names. Could they be made clearer?

Reply:
`vcache -> process_cache` and `rcache -> result_cache`?

Reply:
Or `vcache -> visited_cache` perhaps.

Does `rcache` need to be a dict or can it be a set? In which case it could be `seen_exprs` or `object_cache` or something.

Reply:
Changed `vcache -> visited_cache` and `rcache -> result_cache`.

I think `rcache` wants to be a `dict`. If we have `new_o in rcache`, we know that there is some `o` in `rcache` for which `hash(o) == hash(new_o)` and `o == new_o`, but if `rcache` was just a set, I don't see a straightforward way of getting `o` out of `rcache`.

Reply:
Ah fair enough. It just feels odd to write `cache[result] = result`. But I agree with you now.

And `result_cache` is good.